src/snowflake/snowpark/profiler.py

```diff
@@ -0,0 +1,210 @@
+
+
+class Profiler:
+    """
+    Setup profiler to receive profiles of stored procedures.
+
+    Note:
+        This feature cannot be used in owner's right SP because owner's right SP will not be able to set session-level parameters.
+    """
+
+    def __init__(
+        self,
+        stage: Optional[str] = "",
+        active_profiler: Optional[str] = "LINE",
+        session: Optional["snowflake.snowpark.Session"] = None,
+    ) -> None:
+        self.stage = stage
+        self.active_profiler = active_profiler
+        self.modules_to_register = []
+        self.register_modules_sql = ""
+        self.set_targeted_stage_sql = ""
+        self.enable_profiler_sql = ""
+        self.disable_profiler_sql = ""
+        self.set_active_profiler_sql = ""
+        self.pattern = r"WITH\s+.*?\s+AS\s+PROCEDURE\s+.*?\s+CALL\s+.*"
+        self.session = session
+        self._prepare_sql()
+        self.query_history = None
+
+    def _prepare_sql(self):
+        self.register_modules_sql = f"alter session set python_profiler_modules='{','.join(self.modules_to_register)}'"
+        self.set_targeted_stage_sql = (
+            f'alter session set PYTHON_PROFILER_TARGET_STAGE ="{self.stage}"'
+        )
+        self.enable_profiler_sql = "alter session set ENABLE_PYTHON_PROFILER = true"
+        self.disable_profiler_sql = "alter session set ENABLE_PYTHON_PROFILER = false"
+        self.set_active_profiler_sql = f"alter session set ACTIVE_PYTHON_PROFILER = '{self.active_profiler.upper()}'"
+
+    def register_profiler_modules(self, modules: List[str]):
+        """
+        Register stored procedures to generate profiles for them.
+
+        Note:
+            Registered nodules will be overwritten by this function,
+            use this function with an empty string will remove registered modules.
+        Args:
+            modules: List of names of stored procedures.
+        """
+        self.modules_to_register = modules
+        self._prepare_sql()
+        if self.session is not None:
+            self._register_modules()
+
+    def set_targeted_stage(self, stage: str):
+        """
+        Set targeted stage for profiler output.
+
+        Note:
+            The stage name must be a fully qualified name.
+
+        Args:
+            stage: String of fully qualified name of targeted stage
+        """
+        validate_object_name(stage)
+        self.stage = stage
+        self._prepare_sql()
+        if self.session is not None:
+            if (
+                len(self.session.sql(f"show stages like '{self.stage}'").collect()) == 0
+                and len(
+                    self.session.sql(
+                        f"show stages like '{self.stage.split('.')[-1]}'"
+                    ).collect()
+                )
+                == 0
+            ):
+                self.session.sql(
+                    f"create temp stage if not exists {self.stage} FILE_FORMAT = (RECORD_DELIMITER = NONE FIELD_DELIMITER = NONE )"
```

Comment:
temp stage gets dropped at the end of the session which means we might lose all the profiler data.
when are users going to access the file, after the session or during the session?

Reply:
If user want to access it after the session, user would want to use a permanent stage they created.
when user give a stage name that does not exist, we expect user want to create a temp stage.